### ISSUE TITLE
[HOLD for payment 2024-06-06] [$500]  Web - CTRL + K does not open Search while user avatar preview opened

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#37634]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.46-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4355697
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
1. Log in to New Expensify
2. Navigate to any chat
3. Click on user name in chat history
4. Click on user avatar in RHP
5. Press CTRL/CMD + K on keyboard
6. Verify Search is opened
7. Click on FAB > Start chat

## Expected Result:
Avatar preview should close, and left-hand Search page should open when CTRL + K pressed

## Actual Result:
Avatar preview closes, but the search pages does not open until user opens any RHP e.g. Start chat

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/67a8ff99-b8f7-4a6e-9429-ab3452220b52

 
[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0165d96894b5f57808</li>
        <li>Upwork Job ID: 1770923450252546048</li>
        <li>Last Price Increase: 2024-03-28</li>
<li>Automatic offers: </li>
<ul>
<li>tienifr | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @allroundexperts / @slafortune</details>

**Comment by lanitochka17** [comment]
@slafortune  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave8
CC @zanyrenney

**Comment by slafortune** [comment]
I'm not able to recreate this, are you able to @lanitochka17 ?

**Comment by slafortune** [comment]
Closing - not able to recreate

**Comment by allroundexperts** [comment]
@slafortune This is still reproducible. @bernhardoj has a proposal that might fix it. Can you please re-open this?

**Comment by slafortune** [comment]
You bet - Thanks @allroundexperts

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Can't open the search page using the keyboard shortcut while on a workspace avatar page.

### What is the root cause of that problem?
When we press the search shortcut, the navigation code is wrapped with `Modal.close`.
https://github.com/Expensify/App/blob/7e4a4383a3b7474a15aaba210711a618cd897cbb/src/libs/Navigation/AppNavigator/AuthScreens.tsx#L231-L235

`Modal.close` will call the callback immediately if there is no modal visible (closeModals.length === 0), otherwise, it will store the callback to the "pending" close (onModalClose), close the visible modal, and wait until the modal is hidden.
https://github.com/Expensify/App/blob/7e4a4383a3b7474a15aaba210711a618cd897cbb/src/libs/actions/Modal.ts#L42-L50

In our case, `Modal.close` will close the user avatar modal first. When a modal is hidden, it is supposed to call `onModalDidClose` which will call the pending callback and execute the navigation to the search page.
https://github.com/Expensify/App/blob/7e4a4383a3b7474a15aaba210711a618cd897cbb/src/libs/actions/Modal.ts#L52-L63

`onModalDidClose` itself is called in the Modal `hideModal` function, but `hideModal` is never called.
https://github.com/Expensify/App/blob/575d0ff04101e0690fec251abe75049727d8d8d9/src/components/Modal/BaseModal.tsx#L63-L72

`hideModal` will either be called when the avatar modal unmounts,
https://github.com/Expensify/App/blob/22cb01c968a53407f7a9d864bfba30f41366f6a0/src/components/Modal/BaseModal.tsx#L96-L106

or when `onModalHide` is called.
https://github.com/Expensify/App/blob/22cb01c968a53407f7a9d864bfba30f41366f6a0/src/components/Modal/BaseModal.tsx#L187

When we call `Modal.close`, it will call the AttachmentModal `closeModal` which sets the visibility to false and calls `onModalClose`.
https://github.com/Expensify/App/blob/edfb3b19881b640e1c8dc78f9f07b25ba446c19e/src/components/AttachmentModal.tsx#L391-L396

And for the avatar modal, we navigate back in `onModalClose`.
https://github.com/Expensify/App/blob/edfb3b19881b640e1c8dc78f9f07b25ba446c19e/src/pages/settings/Profile/ProfileAvatar.tsx#L42-L44

So, `onModalHide` didn't get a chance to be called because the modal is unmounted earlier before the modal is completely hidden.

The `hideModal` also failed to be called here because the visible state is already false.
https://github.com/Expensify/App/blob/575d0ff04101e0690fec251abe75049727d8d8d9/src/components/Modal/BaseModal.tsx#L97-L107

### What changes do you think we should make in order to solve the problem?

We need to make sure the `hideModal` is called if the Modal is unmounted and `hideModal` is never called yet. To achieve that, we can create a new ref called `shouldCallHideModalOnUnmount`.

```
const shouldCallHideModalOnUnmount = useRef(false);
...
// unmount code
if (!shouldCallHideModalOnUnmount.current) {
    return;
}
hideModal(true);
```

We set it to true when the modal becomes visible.
https://github.com/Expensify/App/blob/ddf9e7ae25fcc9b0a2516ed341947ad38130b36f/src/components/Modal/BaseModal.tsx#L83-L87

And we set it to false when `hideModal` is called.
https://github.com/Expensify/App/blob/ddf9e7ae25fcc9b0a2516ed341947ad38130b36f/src/components/Modal/BaseModal.tsx#L63-L65

This will make sure the `hideModal` is called in a case where the Modal is previously visible and unmounted, but doesn't have the chance to call `hideModal`.

**Comment by slafortune** [comment]
Checking with @allroundexperts [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1711639748514259)

**Comment by slafortune** [comment]
I've assigned you @allroundexperts 👍

**Comment by allroundexperts** [comment]
Thanks for the proposals @tienifr and @bernhardoj! 

This is definitely a little tricky since changing the base modal introduces regressions very easily. 

@tienifr Can you please double check what issue caused us to add this current `isVisibleRef` logic, and if applying your changes won't recreate that issue again?

I think @bernhardoj's proposal would work as well, while maintaining the previous code. However, I think it would be better if we can fix the previous ref instead of adding a new one.

**Comment by allroundexperts** [comment]
@bernhardoj Since you've thoroughly researched on this issue, do you see any problems with @tienifr's approach?

**Comment by bernhardoj** [comment]
@tienifr approach is basically the same as mine, 
- I changed the name from `isVisibleRef` to `shouldCallHideModalOnUnmount`.
- I set it to true when `isVisible` becomes true, @tienifr set it to true in `onModalShow`
- Both set to false in `hideModal` (`onModalHide`).

So, I'm not adding a new ref but replacing the `isVisibleRef` (I missed pointing this out on the proposal). `isVisibleRef` is added in this [PR](https://github.com/Expensify/App/pull/27639) that I authored to fix this issue https://github.com/Expensify/App/issues/27359#issuecomment-1720990255. The root cause is on the 2nd point of the proposal. To explain it in short, the issue is that the `hideModal` is called twice, one from `onModalHide`, and the other one from when the component is unmounted. We already prevent `hideModal` from being called by checking the `isVisible` state, but since the `useEffect` has an empty deps, the state is never updated, so `isVisible` is always `false` and then I decided to use ref for that.

But now we're having an issue where `hideModal` isn't called at all. So the solution is to make sure `hideModal` is called when unmounted when it's previously visible and `hideModal` is never called yet (`shouldCallHideModalOnUnmount` is true). That's why I renamed it because the ref wouldn't reflect the visible state anymore.

**Comment by bernhardoj** [comment]
You can read in https://github.com/Expensify/App/issues/37609#issuecomment-2003052793 that I already mentioned to change the ref.

**Comment by slafortune** [comment]
@allroundexperts will you have a chance to review this?

**Comment by allroundexperts** [comment]
That works for me. @tienifr can you follow up on this, please?

**Comment by aldo-expensify** [comment]
@tienifr do you have some time to open an issue and ask about if `onModalHide` should be called if the modal gets unmounted? If you don't I think I'll just @bernhardoj 's proposal

**Comment by aldo-expensify** [comment]
I notice there is this `onDismiss` in `ModalContent`: https://github.com/Expensify/App/blob/828c85f4cd9e45971ea5fb0ddf6e86de127fe4dc/src/components/Modal/ModalContent.tsx#L9-L21

which seems to be called if the content is unmounted too. Can that help?

**Comment by bernhardoj** [comment]
From my testing, if we call `hideModal` in `onDismiss` and we close the modal, `hideModal` will be called multiple times. Once from `onModalHide`
https://github.com/Expensify/App/blob/b49288eb7db8cee61a9ab826b8db1e623d1a7460/src/components/Modal/BaseModal.tsx#L210

and twice (idk why) from `onDismiss`.

**Comment by aldo-expensify** [comment]
Waiting on @tienifr , no overdue

**Comment by aldo-expensify** [comment]
Thanks @tienifr

**Comment by aldo-expensify** [comment]
Thanks @tienifr !

**Comment by slafortune** [comment]
Still on hold

**Comment by slafortune** [comment]
Still on hold

**Comment by trjExpensify** [comment]
There was a response saying the maintainers of that upstream library haven't done anything in years, and it does look like there hasn't been any PRs merged for years. What's our alternative here, can we use a patch?

**Comment by aldo-expensify** [comment]
> What's our alternative here, can we use a patch?

We could I guess, but it may be harder to maintain compared to fixing this in our side. I'm not sure what is best really.

**Comment by allroundexperts** [comment]
I would prefer a patch. Whenever we'll update library, we'll be on the look for a fix.

**Comment by trjExpensify** [comment]
Shall we do that then? Otherwise we'll be waiting forever.

**Comment by trjExpensify** [comment]
Go for it!

**Comment by bernhardoj** [comment]
@aldo-expensify @allroundexperts hmm, shouldn't I be the one that gets assigned if we are going to fix this on the App based on this [comment](https://github.com/Expensify/App/issues/37634#issuecomment-2050374691)?

**Comment by slafortune** [comment]
I will be on vacation until June 4th, so adding another BZ in case there is a need.

**Comment by slafortune** [comment]
Changing back to weekly since this is under review, assigning myself back as well to take this back on the 4th.  TIA @puneetlath

**Comment by puneetlath** [comment]
When you say it's under review, what do you mean @slafortune? Is it not on hold anymore?

**Comment by slafortune** [comment]
Whoops!  I should have updated the title - correct!  Not on hold and there is a PR being reviewed.

**Comment by slafortune** [comment]
@allroundexperts can you please complete the checklist?

**Comment by aldo-expensify** [comment]
Friendly bump @allroundexperts

**Comment by allroundexperts** [comment]
## Checklist

1. This was an issue on the upstream repo. As such, there isn't any culprit PR.
2. N/A
3. N/A
4. A regression test would be helpful here. The steps given in the OP look good enough to me.

**Comment by slafortune** [comment]
1. Log in to New Expensify
2. Navigate to any chat
3. Click on user name in chat history
4. Click on user avatar in RHP
5. Press CTRL/CMD + K on keyboard
6. Verify Search is opened
7. Click on FAB > Start chat

**Comment by slafortune** [comment]
@allroundexperts C+ is due $500 via NewDot

**Comment by JmillsExpensify** [comment]
$500 approved for @allroundexperts
